### PR TITLE
RST-3950: Add timeout to handshake

### PR DESCRIPTION
**Pull request by MJohnson459**
Set a socket timeout when waiting to `recv()` the handshake header. For some reason this is done in the main "accept" thread so this will block everything if there isn't a timeout. I don't know why the handshake isn't being sent or received.

Setting a timeout of 1 second seems to work well in the sim but even then with 200 robots restarting I had about 25 of them hitting the timeout. So even with a relatively short timeout the accept thread is still blocked for 25 seconds...

---

I tracked this down using the stack traces of a working instances threads and one that was locked up. Most of the traces were identical apart from this one:

working:
```
Thread 3454 (idle): "Thread-4"
    accept (socket.py:205)
    run (rospy/impl/tcpros_base.py:154)
    run (threading.py:864)
    _bootstrap_inner (threading.py:916)
    _bootstrap (threading.py:884)
```

locked:
```
Thread 11282 (idle): "Thread-4"
    read_ros_handshake_header (rosgraph/network.py:357)
    _tcp_server_callback (rospy/impl/tcpros_base.py:325)
    run (rospy/impl/tcpros_base.py:167)
    run (threading.py:864)
    _bootstrap_inner (threading.py:916)
    _bootstrap (threading.py:884)
```

### CODE REVIEW

**Review by paulbovbel — APPROVED** [review]
thanks! this looks great. I'm going to test this with my yammer sim

File: tools/rosgraph/src/rosgraph/network.py [hunk 1]
```diff
@@ -354,9 +354,19 @@ def read_ros_handshake_header(sock, b, buff_size):
...
+            sock.settimeout(1.0)
+            d = sock.recv(buff_size)
+        except socket.timeout:
+            raise ROSHandshakeException("connection from sender timed out before handshake header received. Please check sender %s:%d for additional details." % sock.getpeername())
+        finally:
+            sock.settimeout(prev_timeout)
+
         if not d:
             raise ROSHandshakeException("connection from sender terminated before handshake header received. %s bytes were received. Please check sender for additional details."%b.tell())
+
```
Comment (ayrton04) [inline tools/rosgraph/src/rosgraph/network.py]: Just because it's a `ros_comm` PR, we should limit unnecessary changes.

**Comment by ayrton04** [comment]
If comms are really poor at the site, is it possible we'd never receive the handshake header within 1 second? Is there every a benefit to a longer timeout?

**Review by ayrton04 — APPROVED** [review]

**Comment by MJohnson459** [comment]
> If comms are really poor at the site, is it possible we'd never receive the handshake header within 1 second? Is there every a benefit to a longer timeout?

There are two situations where the timeout will occur - when the header just isn't sent and when a robot has a bad connection. 
In the first case the timeout will always freeze the node for the full timeout length which in a test case was roughly 10% of the bots - about 25 bots in 200 and freezing for 25s isn't terrible. Even 50s (2s timeout) or 75s (3s timeout) isn't catastrophic.

In the case of a bad connection, we might just never receive the header. This wouldn't be so bad if the robot was moving as the connection might improve but a lot of the time the issue will happen is when the robot is sitting still waiting for a queue position or having just been turned on.

In conclusion, it's probably safest to increase the timeout to about 3s. If we turn on all the bots at once (which is really the worst case) then we would expect it to take a couple of minutes to have all the robots get the latest site data but it won't lock the node which is an improvement. If a robot has bad comms then it can have a ping of up to 3s before we start seeing issues - if it has more than that then its probably stuck anyway.

**Review by ayrton04 — APPROVED** [review]
